File: android/app/src/main/java/info/bagen/dwebbrowser/microService/browser/jmm/JmmMetadata.kt
package info.bagen.dwebbrowser.microService.browser.jmm

import org.dweb_browser.microservice.help.DWEB_DEEPLINK
import org.dweb_browser.microservice.help.Mmid

data class JmmMetadata(
  val id: Mmid, // jmmApp的id
  val server: MainServer, // 打开应用地址
  val dweb_deeplinks: MutableList<DWEB_DEEPLINK> = mutableListOf(),
  val name: String = "", // 应用名称
  val short_name: String = "", // 应用副标题
  val icon: String = "", // 应用图标
  val images: List<String>? = null, // 应用截图
  val description: String = "", // 应用描述
  val author: List<String>? = null, // 开发者，作者
  val categories: List<String>? = null, // 应用类型 https://github.com/w3c/manifest/wiki/Categories
  val version: String = "", // 应用版本
  val new_feature: String? = null, // 新特性，新功能
  val home: String = "", // 首页地址
  var bundle_url: String = "", // 下载应用地址
  val bundle_size: String = "", // 应用大小
  val bundle_hash: String = "", // 文件hash
  val permissions: List<String>? = null, // app使用权限的情况
  val plugins: List<String>? = null, // app使用插件的情况
  val release_date: String = "", // 发布时间
) : java.io.Serializable {
    data class MainServer(
        /**
         * 应用文件夹的目录
         */
        val root: String,
        /**
         * 入口文件
         */
        val entry: String
    ) : java.io.Serializable


}


val defaultJmmMetadata = JmmMetadata(
    id = "demo.user.dweb",
    dweb_deeplinks = mutableListOf(),
    icon = "https://www.bfmeta.info/imgs/logo3.webp",
    name = "plaoc demo",
    short_name = "demo",
    description = "这是一个实例应用，包含了dweb_plugins全部组件的实例。",
    bundle_url = "https://dweb.waterbang.top/game.dweb.waterbang.top.dweb-1.0.0.zip",
    bundle_size = "2742079",
    bundle_hash = "sha256:966251757d12a7f4021a0b40031ff4b2b60ae08b63de516b572eb572bfd90886", // 文件hash
    version = "1.2.0",
    images = listOf(
        "http://qiniu-waterbang.waterbang.top/bfm/cot-home_2058.webp",
        "http://qiniu-waterbang.waterbang.top/bfm/defi.png",
        "http://qiniu-waterbang.waterbang.top/bfm/nft.png"
    ),
    home = "https://dweb-browser.org/",
    server = JmmMetadata.MainServer(
        root = "/sys",
        entry = "/server/plaoc.server.js"
    ),
    author = listOf("bfs", "bfs@bfs.com"),
    categories = listOf("demo", "vue3","plaoc"),
    release_date = "2023-05-22T09:01:38.318Z",
)

File: next/dweb-browser/src/Platforms/iOS/MicroModule/Browser/Jmm/JmmMetadata.cs
using System.Text.Json;
using System.Text.Json.Serialization;

#nullable enable

namespace DwebBrowser.MicroService.Browser.Jmm;

public class JmmMetadata
{
    [JsonPropertyName("id")]
    public Mmid Id { get; set; }    // jmmApp的id
    [JsonPropertyName("server")]
    public MainServer Server { get; set; }      // 打开应用地址
    [JsonPropertyName("dweb_deeplinks")]
    public List<Dweb_DeepLink>? Dweb_DeepLinks { get; set; }       // dweb-deeplinks
    [JsonPropertyName("name")]
    public string Name { get; set; }       // 应用名称
    [JsonPropertyName("short_name")]
    public string ShortName { get; set; }        // 应用副标题
    [JsonPropertyName("icon")]
    public string Icon { get; set; }        // 应用图标
    [JsonPropertyName("images")]
    public List<string>? Images { get; set; }       // 应用截图
    [JsonPropertyName("description")]
    public string Description { get; set; }        // 应用描述
    [JsonPropertyName("author")]
    public List<string>? Author { get; set; }       // 开发者，作者
    [JsonPropertyName("version")]
    public string Version { get; set; }     // 应用版本
    [JsonPropertyName("new_feature")]
    public string NewFeature { get; set; }      // 新特性，新功能
    [JsonPropertyName("categories")]
    public List<string>? Categories { get; set; }     // 关键词
    [JsonPropertyName("home")]
    public string Home { get; set; }        // 首页地址
    [JsonPropertyName("bundle_url")]
    public string BundleUrl { get; set; }     // 下载应用地址
    [JsonPropertyName("bundle_size")]
    public long BundleSize { get; set; }        // 应用大小
    [JsonPropertyName("bundle_hash")]
    public string BundleHash { get; set; }        // 文件hash
    [JsonPropertyName("permissions")]
    public List<string>? Permissions { get; set; }      // app使用权限情况
    [JsonPropertyName("plugins")]
    public List<string>? Plugins { get; set; }      // app使用插件情况
    [JsonPropertyName("release_date")]
    public string ReleaseDate { get; set; }     // 发布时间

    [Obsolete("使用带参数的构造函数", true)]
#pragma warning disable CS8618 // 在退出构造函数时，不可为 null 的字段必须包含非 null 值。请考虑声明为可以为 null。
    public JmmMetadata()
#pragma warning restore CS8618 // 在退出构造函数时，不可为 null 的字段必须包含非 null 值。请考虑声明为可以为 null。
    {
        /// 给JSON反序列化用的空参数构造函数
    }

    public JmmMetadata(
        Mmid id,
        MainServer server,
        List<Dweb_DeepLink>? dweb_deeplinks = null,
        string name = "",
        string shortName = "",
        string icon = "",
        List<string>? images = null,
        string description = "",
        List<string>? author = null,
        string version = "",
        string new_feature = "",
        List<string>? categories = null,
        string home = "",
        string bundle_url = "",
        long bundle_size = 0L,
        string bundle_hash = "",
        List<string>? permissions = null,
        List<string>? plugins = null,
        string release_date = "")
    {
        Id = id;
        Server = server;
        Name = name;
        ShortName = shortName;
        Icon = icon;
        Images = images;
        Description = description;
        Author = author;
        Version = version;
        NewFeature = new_feature;
        Categories = categories;
        Home = home;
        BundleUrl = bundle_url;
        BundleSize = bundle_size;
        BundleHash = bundle_hash;
        Permissions = permissions;
        Plugins = plugins;
        ReleaseDate = release_date;
        if (dweb_deeplinks is null)
        {
            Dweb_DeepLinks = new();
        }
        else
        {
            Dweb_DeepLinks = dweb_deeplinks;
        }
    }

    public struct MainServer
    {
        /// <summary>
        /// 应用文件夹的目录
        /// </summary>
        [JsonPropertyName("root")]
        public string Root { get; set; }

        /// <summary>
        /// 入口文件
        /// </summary>
        [JsonPropertyName("entry")]
        public string Entry { get; set; }
    }

    public string ToJson() => JsonSerializer.Serialize(this, new JsonSerializerOptions { WriteIndented = true });
    public static JmmMetadata? FromJson(string json) =>
        JsonSerializer.Deserialize<JmmMetadata>(json, new JsonSerializerOptions { IncludeFields = true });
}
